cmd/bnsd/x/blueaccount/handler.go

```diff
@@ -0,0 +1,519 @@
+}
+
+func (h *registerDomaiHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	owner := msg.Owner
+	if len(owner) == 0 {
+		owner = x.MainSigner(ctx, h.auth).Address()
+	}
+
+	now, err := weave.BlockTime(ctx)
+	if err != nil {
+		return nil, errors.Wrap(errors.ErrState, "block time not present in context")
+	}
+	conf, err := loadConf(db)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load configuration")
+	}
+	domain := Domain{
+		Metadata:    &weave.Metadata{},
+		Owner:       owner,
+		Domain:      msg.Domain,
+		ValidUntill: weave.AsUnixTime(now.Add(conf.DomainRenew.Duration())),
+	}
+	if _, err := h.domains.Put(db, []byte(msg.Domain), &domain); err != nil {
+		return nil, errors.Wrap(err, "cannot store domain entity")
+	}
+
+	// Registering a domain enforce existence of a username with an empty
+	// name.
+	account := Account{
+		Metadata: &weave.Metadata{},
+		Owner:    nil, // Always delegate to the domain owner, never set explicitly.
+		Domain:   msg.Domain,
+		Name:     "",
+	}
+	if _, err := h.accounts.Put(db, accountKey("", msg.Domain), &account); err != nil {
+		return nil, errors.Wrap(err, "cannot store account entity")
+	}
+
+	return &weave.DeliverResult{Data: nil}, nil
+}
+
+func (h *registerDomaiHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*RegisterDomainMsg, error) {
+	var msg RegisterDomainMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+	switch err := h.domains.Has(db, []byte(msg.Domain)); {
+	case err == nil:
+		return nil, errors.Wrapf(errors.ErrDuplicate, "domain %q already registered", msg.Domain)
+	case errors.ErrNotFound.Is(err):
+		// All good.
+	default:
+		return nil, errors.Wrap(err, "cannot check if domain already exists")
+	}
+
+	conf, err := loadConf(db)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load configuration")
+	}
+	if ok, err := regexp.MatchString(conf.ValidDomain, msg.Domain); err != nil || !ok {
+		return nil, errors.Wrap(errors.ErrInput, "domain is not allowed")
+	}
+
+	return &msg, nil
+}
+
+type transferDomaiHandler struct {
+	auth    x.Authenticator
+	domains orm.ModelBucket
+}
+
+func (h *transferDomaiHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: 0}, nil
+}
+
+func (h *transferDomaiHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	domain, msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+	domain.Owner = msg.NewOwner
+	if _, err := h.domains.Put(db, []byte(msg.Domain), domain); err != nil {
+		return nil, errors.Wrap(err, "cannot store domain")
+	}
+	return &weave.DeliverResult{Data: nil}, nil
```

Comment:
Just a question that will help me learn more. Why use `nil` here? Possibly you can return a value that will help the user to interact with blockchain in the next txs

Reply:
We rarely return something here. It makes sense to return an ID of the entity that was created, so that you can reference it.
In this case ID is provided by the users. `domain` is part of the input, which means that there is no need to return it. 

This case is repeating for an account registration. There ID is constructed from two values  - `domain` and `name`. In this case it cannot even be returned as single `Data`.

Here is an explicit `nil` instead of the usual `&weave.DeliverResult{}, nil`, because this code was generated. I think it is easier to update this code if you know which attribute to change. I can as well remove `Data: nil` part. It makes no difference.